dali/operators/audio/preemphasis_filter_op.cu

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/audio/preemphasis_filter_op.h"
+#include <vector>
+
+namespace dali {
+
+namespace detail {
+
+template <typename OutputType, typename InputType>
+struct SampleDescriptor {
+  const InputType *in;
+  OutputType *out;
+  float coeff;
+  int64_t size;
+};
+
+template <typename OutputType, typename InputType>
+void __global__ PreemphasisFilterKernel(const SampleDescriptor<OutputType, InputType> *samples) {
+  const auto &sample = samples[blockIdx.y];
+  int64_t block_size = blockDim.x;
+  int64_t block_start = block_size * blockIdx.x;
+  int64_t grid_stride = block_size * gridDim.x;
+  for (int64_t k = block_start + threadIdx.x; k < sample.size; k += grid_stride) {
+    auto prev = cuda_max(k-1, 0l);
+    sample.out[k] = sample.in[k] - sample.coeff * sample.in[prev];
+  }
```

Comment:
It should be:
```C++
  int64_t k = block_start + threadIdx.x
  if (k >= sample_size)
    return;

  sample.out[k] = sample.in[k] - sample.coeff  * sample.in[cuda_max(0, k-1)];
  for (; k < sample.size; k += grid_stride) {
    sample.out[k] = sample.in[k] - sample.coeff * sample.in[k-1];
  }
```

Reply:
I thought it was better to avoid conditional branching when writing CUDA kernels to avoid thread divergence.
@mzient OK, I'll limit cuda_max usage to the first iteration only
@awolant Keep in mind that threadIdx.x == 0 && blockIdx.x == 0 doesn't mean start of the sample, since every sample is divided in blocks.

Reply:
As far as I can tell, in the original code prev is not k-1 only for k == 0. And k is a sum of threadIdx.x and (block_size * blockIdx.x), so it is equal to 0 only for threadIdx.x == 0 and blockIdx.x == 0.
I've tested this code and it works with unit tests added in this PR.